DylibBinder/DBGenericParameter.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Generic;
+using Dynamo.SwiftLang;
+using SwiftReflector;
+using SwiftReflector.Inventory;
+
+namespace DylibBinder {
+	internal class DBGenericParameter {
+		public DBGenericParameter (int depth, int index)
+		{
+			Depth = depth;
+			Index = index;
+			Name = SLGenericReferenceType.DefaultNamer (depth, index);
+		}
+
+		public int Depth { get; }
+		public int Index { get; }
+		public string Name { get; }
+	}
+
+	internal class DBGenericParameters {
+
+		public DBGenericParameters (SwiftBaseFunctionType signature) {
+			GenericParameters.AddRange (signature.AssignGenericParameters ());
+		}
+
+		public DBGenericParameters (SwiftType swiftType)
+		{
+			GenericParameters.AddRange (swiftType.AssignGenericParameters ());
+		}
+
+		public DBGenericParameters (DBTypeDeclaration typeDeclaration)
+		{
+			GenericParameters.AddRange (typeDeclaration.ParseTopLevelGenerics ());
+		}
+
+		public HashSet<DBGenericParameter> GenericParameters { get; } = new HashSet<DBGenericParameter> (new DBGenericParameterComparer ());
+
+		public static List<DBGenericParameter> FilterGenericParameterSource (params SwiftType [] types)
+		{
+			var genericParameters = new List<DBGenericParameter> ();
+			foreach (var type in types) {
+				genericParameters.AddRange (GenericParameterSwitch (type));
+			}
+			return genericParameters;
+		}
+
+		public static List<DBGenericParameter> GetGenericParameters (SwiftFunctionType funcType)
+		{
+			var genericParameters = new List<DBGenericParameter> ();
+			genericParameters.AddRange (GenericParameterSwitch (funcType.Parameters));
+			return genericParameters;
+		}
+
+		public static List<DBGenericParameter> GetGenericParameters (SwiftBoundGenericType boundType)
+		{
+			var genericParameters = new List<DBGenericParameter> ();
+			foreach (var bound in boundType.BoundTypes) {
+				if (bound is SwiftGenericArgReferenceType refType)
+					genericParameters.AddRange (GetGenericParameters (refType));
+			}
+			return genericParameters;
+		}
+
+		public static List<DBGenericParameter> GetGenericParameters (SwiftTupleType tupleType)
+		{
+			var genericParameters = new List<DBGenericParameter> ();
+			foreach (var content in tupleType.Contents) {
+				genericParameters.AddRange (GenericParameterSwitch (content));
+			}
+			return genericParameters;
+		}
+
+		public static List<DBGenericParameter> GetGenericParameters (SwiftGenericArgReferenceType refType)
```

Comment:
Would it be okay to make it a private static? I remember that you did not like private static methods before, but my extension methods call the static methods that call these so these need to be static if I am correct?

I believe I need a list because there could be 0-{some number greater than 4} generic parameters found